File: Windows 8.1 app samples/HttpClient sample (Windows 8.1)/C++/HttpClient/PlugInFilter.cpp
#include "pch.h"
#include "PlugInFilter.h"
#include <ppltasks.h>

using namespace Concurrency;
using namespace Platform;
using namespace SDKSample::HttpClientSample;
using namespace Windows::Foundation;
using namespace Windows::Web::Http;
using namespace Windows::Web::Http::Filters;

PlugInFilter::PlugInFilter(IHttpFilter^ innerFilter)
{
    if (innerFilter == nullptr)
    {
        throw ref new Exception(E_INVALIDARG, "innerFilter cannot be null.");
    }
    this->innerFilter = innerFilter;
}

PlugInFilter::~PlugInFilter(void)
{
}

IAsyncOperationWithProgress<HttpResponseMessage^, HttpProgress>^ PlugInFilter::SendRequestAsync(
    HttpRequestMessage^ request)
{
    request->Headers->Insert("Custom-Header", "CustomRequestValue");

    return create_async([=](progress_reporter<HttpProgress> reporter, cancellation_token token){
        HttpResponseMessage^ response = nullptr;

        task<HttpResponseMessage^> sendRequestTask(innerFilter->SendRequestAsync(request), token);
        sendRequestTask.wait();
        response = sendRequestTask.get();

        if (token.is_canceled())
        {
            cancel_current_task();
            return response;
        }

        response->Headers->Insert("Custom-Header", "CustomResponseValue");
        return response;
    });
}

File: Windows 8.1 app samples/HttpClient sample (Windows 8.1)/C#,C++/HttpClient/PlugInFilter.cs
﻿using System;
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Foundation;
using Windows.Web.Http;
using Windows.Web.Http.Filters;

namespace Microsoft.Samples.Networking.HttpClientSample
{
    public class PlugInFilter : IHttpFilter
    {
        private IHttpFilter innerFilter;

        public PlugInFilter(IHttpFilter innerFilter)
        {
            if (innerFilter == null)
            {
                throw new ArgumentException("innerFilter cannot be null.");
            }
            this.innerFilter = innerFilter;
        }

        public IAsyncOperationWithProgress<HttpResponseMessage, HttpProgress> SendRequestAsync(HttpRequestMessage request)
        {
            return AsyncInfo.Run<HttpResponseMessage, HttpProgress>(async (cancellationToken, progress) =>
            {
                request.Headers.Add("Custom-Header", "CustomRequestValue");
                HttpResponseMessage response = await innerFilter.SendRequestAsync(request).AsTask(cancellationToken, progress);

                cancellationToken.ThrowIfCancellationRequested();

                response.Headers.Add("Custom-Header", "CustomResponseValue");
                return response;
            });
        }

        public void Dispose()
        {
            innerFilter.Dispose();
            GC.SuppressFinalize(this);
        }
    }
}
